Fix Prefix Matching Logic in PrefixContainer

sami-daniel opened the pull request:
# Fix Prefix Matching Logic in PrefixContainer

## Description

This PR modifies the `BinarySearch` method in `PrefixContainer` to ensure that an exact match of a prefix does not incorrectly qualify as a valid prefix.

Now the ContainsPrefix method no longer gives a false positive if it has a prefix with the same name as the Action method parameter.

Fixes #57637

Commit:
Refactor PrefixContainerTest to include a new test case for checking if the prefix matches exactly with keys

Commit:
Fix BinarySearch logic

Updated the BinarySearch method in PrefixContainer to ensure that an exact match of a prefix does not incorrectly qualify as a valid prefix. Enhanced comments to clarify the flow and decision-making process when checking for delimiters. This change improves the accuracy of prefix matching for hierarchical key structures.

Commit:
Fix the name of the test that checks the Return False of ContaisPrefix if the Prefix is ​​equal to the keys

sami-daniel commented:
> @sami-daniel please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

gfoidl commented on src/Mvc/Mvc.Core/src/ModelBinding/PrefixContainer.cs:
```diff
@@ -201,32 +201,40 @@ private int BinarySearch(string prefix)
+                // which should not qualify as a prefix match. Therefore, we will continue searching
+                // for valid matches in this case.
+
+                if (candidate.Length > prefix.Length)
```
In order to safely access `candidate[prefix.Length]` the JIT emits a bound check. This bound check can be elided by 
```suggestion
                if ((uint)candidate.Length > (uint)prefix.Length)
```
See [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAMQgQAFMXIAGBmGwCAJnyXYMMNE1kMADrABmfBFsHDdfAG4QMASi49O1Hi4Z8DDCQuWr1MAHQAMjACAOYYABYMAHx6hsZBIeER9s6u3E7p6WAR2FDyDAC88ooqahoA2vowRgiJYZEAugDcDlk8be1uHl5FxQDk/v0MhIQFhQMV/aldjp2z3MQA7HpWNq1pswC+8zw7m667iytw5Bsu+y6d9G5CDABCeVI63mV+WtJy1bWmdxbWdk6mSy7k8EgArmZbK9fBp6skYp5IUJbN8EsEGikjhxsS4cnlxiUfOUYFV4nUMckWriOgd2qDehMGINhqNxpNpjSMlzXMtVgDzgtLllhelsXzToLuPstkA==) for the difference in codegen.

gfoidl reviewed (commented):
A small JIT-trick for better machine code.

sami-daniel committed:
Optimize Boundary Check in BinarySearch Method

This commit enhances the performance of the BinarySearch method by eliminating unnecessary boundary checks. The condition if ((uint)candidate.Length > (uint)prefix.Length) ensures that the index access candidate[prefix.Length] is safe, and the JIT compiler can now safely elide the boundary check, improving execution efficiency.

Co-authored-by: Günther Foidl <gue@korporal.at>

sami-daniel commented:
> A small JIT-trick for better machine code.

I didn't know that, but it makes sense, thanks!

gfoidl commented:
> I didn't know that

The nice thing on contributing: you help to improve the product and get the opportunity to learn something new. A win - win situation 😃.

Thanks for the PR 👍🏻

sami-daniel commented:
> > I didn't know that
> 
> The nice thing on contributing: you help to improve the product and get the opportunity to learn something new. A win - win situation 😃.
> 
> Thanks for the PR 👍🏻

Yessss, C# and .NET have been my focus of study since the beginning of my interest in technology and until now I don't even know half of its power. This is my first contribution and I already loved it. Thanks!

adityamandaleeka commented:
> Yessss, C# and .NET have been my focus of study since the beginning of my interest in technology and until now I don't even know half of its power. This is my first contribution and I already loved it. Thanks!

Love it, welcome @sami-daniel!

adityamandaleeka commented:
/azp run

adityamandaleeka commented:
Ubuntu failure is unrelated; it failed during node installation. cc: @dotnet/aspnet-build 

```
##[error]Request timeout: /dist/index.json
```

wtgodbe commented:
/azp run

adityamandaleeka commented:
Not sure if there's something making our CI more flaky than normal...

/azp run

sami-daniel commented:
Hello again :).
Would it be possible to run the pipelines gain to see if the behavior is the same? I will make some changes if necessary

wtgodbe commented:
/azp run

sami-daniel committed:
Merge branch 'dotnet:main' into main

captainsafia reviewed (commented):
Thanks for opening this PR!

It looks like the test failures that were introduced here are legitimate failures. Specifically, the failures in `Mvc.FunctionalTests` can be debugged to investigate what regression was introduced here.
